docs: Add documentation around reference frame transforms in Monty

vkakerbeck opened the pull request:
This PR adds documentation around the reference frames document we discussed in past research meetings into the documentation. I added a bit more structure & code references to it since then.

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+---
+title: Reference Frame Transforms
+---
+In a sensorimotor learning setup one naturally encounters several different reference frames in which information can be represented. Additionally, Monty has it's internal reference frames in which is learns models of objects. Those can get confusing to wrap your head around and keep track of so here is a brief overview of all the reference frames involved in a typical Monty setup.
```
todo: commas, typo.

```suggestion
In a sensorimotor learning setup, one naturally encounters several different reference frames in which information can be represented. Additionally, Monty has internal reference frames in which it learns models of objects. Those can get confusing to wrap your head around and keep track of, so here is a brief overview of all the reference frames involved in a typical Monty setup.
```

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+![](../../figures/how-monty-works/reference_frames_overview.png)
+
+### 🟢 Object rel. World
+The orientation of the object in the world (unknown to Monty). For instance, if we use a simulator, this would be the configuration that specifies where the object is placed in the environment (like we do [here](https://github.com/thousandbrainsproject/tbp.monty/blob/4844ef17a4cadce455acb8d852fe3ed7038a298f/src/tbp/monty/frameworks/config_utils/make_dataset_configs.py#L229)). In the real world we ...
```
suggestion: unknown implies it can be learned.  Maybe 'withheld' which indicates its never sent to monty?



```suggestion
The orientation of the object in the world (withheld from Monty). For instance, if we use a simulator, this would be the configuration that specifies where the object is placed in the environment (like we do [here](https://github.com/thousandbrainsproject/tbp.monty/blob/4844ef17a4cadce455acb8d852fe3ed7038a298f/src/tbp/monty/frameworks/config_utils/make_dataset_configs.py#L229)). In the real world we don't have a specific origin but still, all objects are placed in a common reference frame with relative displacements between each other.
```

vkakerbeck replied:
hm technically Monty can infer this (it just doesn't know the ground truth). Do you feel strongly about this? (mostly asking bc I would also have to update the animations & image with that wording but if it is confusing how it is I can do that)

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+### 🟢 Object rel. World
+The orientation of the object in the world (unknown to Monty). For instance, if we use a simulator, this would be the configuration that specifies where the object is placed in the environment (like we do [here](https://github.com/thousandbrainsproject/tbp.monty/blob/4844ef17a4cadce455acb8d852fe3ed7038a298f/src/tbp/monty/frameworks/config_utils/make_dataset_configs.py#L229)). In the real world we ...
+### 🟢⚫️ Feature rel. Object
+The pose of all the features on the object (also unknown to the Monty). For instance, if you use a 3D simulator, this could be defined in the object meshes.
```
suggestion: withheld


```suggestion
The pose of all the features on the object (also withheld from Monty). For instance, if you use a 3D simulator, this could be defined in the object meshes.
```

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+### 🟢⚫️ Feature rel. Object
+The pose of all the features on the object (also unknown to the Monty). For instance, if you use a 3D simulator, this could be defined in the object meshes.
+### 🟡 Sensor rel. World/Body
+The sensor’s location and orientation in the world can be calculated from proprioceptive information and motor efference copies. Basically, as the system moves it's sensors, it can use this information to update the locations of those sensors in a common reference frame. For the purposes of Monty it doesn't matter where the origin of this RF is (it could be the agent's body or an arbitrary locati ...
```
todo: apostrophe.

```suggestion
The sensor’s location and orientation in the world can be calculated from proprioceptive information and motor efference copies. Basically, as the system moves its sensors, it can use this information to update the locations of those sensors in a common reference frame. For the purposes of Monty it doesn't matter where the origin of this RF is (it could be the agent's body or an arbitrary location in the environment) but it matters that all sensor locations are represented in this common RF.
```

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+### 🟡 Sensor rel. World/Body
+The sensor’s location and orientation in the world can be calculated from proprioceptive information and motor efference copies. Basically, as the system moves it's sensors, it can use this information to update the locations of those sensors in a common reference frame. For the purposes of Monty it doesn't matter where the origin of this RF is (it could be the agent's body or an arbitrary locati ...
+### 🟠 Feature rel. Sensor
+This is the feature orientation relative to the sensor. For instance in Monty, we often use the surface normal and curvature direction to define the sensed pose, which are extracted from the depth image of the sensor ([code](https://github.com/thousandbrainsproject/tbp.monty/blob/main/src/tbp/monty/frameworks/models/sensor_modules.py#L161-L167)).
```
question: should this be "the feature's orientation"?

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+This is the feature orientation relative to the sensor. For instance in Monty, we often use the surface normal and curvature direction to define the sensed pose, which are extracted from the depth image of the sensor ([code](https://github.com/thousandbrainsproject/tbp.monty/blob/main/src/tbp/monty/frameworks/models/sensor_modules.py#L161-L167)).
+### 🟢⚪️ Feature rel. World/Body
+[bright green]
+The estimated orientation of the sensed feature in the world (sensor_rel_world * feature_rel_sensor). In Monty this currently happens to the depth values in [DepthTo3DLocations](https://github.com/thousandbrainsproject/tbp.monty/blob/4844ef17a4cadce455acb8d852fe3ed7038a298f/src/tbp/monty/frameworks/environment_utils/transforms.py#L220) while the surface normal and curvature extraction then happen ...
```
todo: comma 

```suggestion
The estimated orientation of the sensed feature in the world (sensor_rel_world * feature_rel_sensor). In Monty, this currently happens to the depth values in [DepthTo3DLocations](https://github.com/thousandbrainsproject/tbp.monty/blob/4844ef17a4cadce455acb8d852fe3ed7038a298f/src/tbp/monty/frameworks/environment_utils/transforms.py#L220) while the surface normal and curvature extraction then happens in the SM, but in the end you get the same results.
```

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+### 🔵 Object rel. Model
+This is the hypothesized orientation of the learned model relative to the object (in the world). This orientation needs to be inferred by the LM based on its sensory inputs. There are usually multiple hypotheses which Monty learning modules keeps track of in [`self.possible_poses`](https://github.com/thousandbrainsproject/tbp.monty/blob/4844ef17a4cadce455acb8d852fe3ed7038a298f/src/tbp/monty/frame ...
+### 🔵⚫️ Feature rel. Model
+This is the rotation of the currently sensed feature relative to the object model in the LM (feature_rel_world * object_rel_model, see code [here](https://github.com/thousandbrainsproject/tbp.monty/blob/main/src/tbp/monty/frameworks/models/evidence_matching/hypotheses_displacer.py#L141-L142)). The learning module uses it's pose hypothesis to transform the feature relative to the world into it's o ...
```
todo: apostrophes

```suggestion
This is the rotation of the currently sensed feature relative to the object model in the LM (feature_rel_world * object_rel_model, see code [here](https://github.com/thousandbrainsproject/tbp.monty/blob/main/src/tbp/monty/frameworks/models/evidence_matching/hypotheses_displacer.py#L141-L142)). The learning module uses its pose hypothesis to transform the feature relative to the world into its object's reference frame so that it can recognize the object in any orientation and location in the world, independent of where it was learned. 
```

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+What is not shown in this image is that the same rotation transform is also applied to the movement vector (displacement) that is applied to update the hypothesized location on the object.
+
+# Reference Frame Transforms for Voting
+Voting in Monty happens in object space. We directly translate between the object’s RF of the sending LM to the object RF of the receiving LM. This relies on the assumption that both LMs learned the object at the same time, and hence their reference frames line up in orientation and displacement (since we receive features rel. world, which will automatically line up if the LMs learn the object at ...
```
suggestion: slight rephrasing

```suggestion
Voting in Monty happens in object space. We directly translate between the object’s RF of the sending LM to the object RF of the receiving LM. This relies on the assumption that both LMs learned the object at the same time, and hence their reference frames line up in orientation and displacement (since we receive features rel. world, which will automatically line up if the LMs learn the object at the same time). Otherwise, we could store one displacement between their reference frames and apply that as well.
```

codeallthethingz commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,56 @@
+
+The two LMs receive input from two sensors that sense different locations and orientations in space. They receive that as a pose in a common coordinate system (rel. world in the image below). Since the sensors are at different locations on the object and our hypotheses are “locations of sensor rel. model” we can’t just vote on the hypotheses directly but have to account for the relative sensor di ...
+
+This offset can be easily calculated from the difference of the two LM’s inputs as those poses are in a common coordinate system. The sending LM attaches its sensed pose in the world to the vote message it sends out (along with its hypotheses about the locations on the mug) and the receiving LM compares it with its own sensed pose and applies the difference to the vote hypotheses.
```
todo: commas

```suggestion
This offset can be easily calculated from the difference of the two LM’s inputs, as those poses are in a common coordinate system. The sending LM attaches its sensed pose in the world to the vote message it sends out (along with its hypotheses about the locations on the mug), and the receiving LM compares it with its own sensed pose and applies the difference to the vote hypotheses.
```

codeallthethingz commented on docs/overview/further-reading.md:
```diff
@@ -303,4 +303,4 @@ Below is a great introductory video of Jeff Hawkins presenting an overview of ou
 }
-[/block]
\ No newline at end of file
+[/block]
```
suggestion: revert this file as it was not changed.

codeallthethingz reviewed (changes requested):
This is such a great reference, the animations are 🔥 !

A few minor typos / grammar suggestions.

vkakerbeck commented:
@codeallthethingz I added some more visualizations and text based on our discussion yesterday. Lmk if you think this is helpful of if anything is missing.

@nielsleadholm this is ready for review now.

Pretty view is here: https://thousandbrainsproject.readme.io/v0.10-vkakerbeck-rf-transform-docs/docs/reference-frame-transforms

codeallthethingz commented:
> @codeallthethingz I added some more visualizations and text based on our discussion yesterday. Lmk if you think this is helpful of if anything is missing.

looks good! 👍

vkakerbeck commented on docs/how-monty-works/monty/reference-frame-transformations.md:
```diff
@@ -0,0 +1,83 @@
+
+![](../../figures/how-monty-works/vote_loc_rot_transform.png#width=400px)
+
+After this displacement is applied, the vote locations (red dots) can again be overlaid onto the existing hypothesis space (orange dots). In this example, the right LM received a pose and features that were consistent with the top or bottom of the handle. After incorporating the votes, the top of the handle is the most likely location (circled in green).
```
Yes, that is correct how you describe it @scottcanoe Maybe one thing to highlight is that the vote doesn't contain any kind of labels like "rim" or "handle", I just refer to those in the text to make it easier to identify the different locations I am talking about. 
The votes themselves only contain locations. Those locations are unique to a specific object (in Monty, the vote dictionary identifies the object ID that the location votes are associated with). 
We work on the assumption that the object was learned in both columns at the same time and that therefore the RFs correspond to each other. So if you look at example 1, where both sensors are sensing the same location in the world, no displacement needs to be applied, and the location votes from LM1 can be directly applied to LM2's RF. If there is a sensor displacement, then the voted locations need to be shifted by that amount. 
If we didn't do this, then the LM that currently thinks it is on the rim would also add evidence for the locations on the rim to LM2, but that LM is sensing a different location on the mug, so it should not expect to be on the rim. Again, "rim" and "handle" are never explicitly part of the votes, but the locations that are sent to LM2 are displaced such that they end up on those locations on the model in this example. Does that make sense?

tristanls replied:
Thank you both.

FWIW, I didn't understand @scottcanoe's description, but I understood @vkakerbeck's "votes only contain locations." I think my sticking point in @scottcanoe's description was "observation", which I interpret differently from "location," since observations are features at poses.

So, translating for my understanding: 

Setup: 
- LM 1 perceives object P at location A
- LM 2 perceives object P at location B

Dynamics:
Learning Modules only vote on locations. When a Learning Module sends out a vote, the vote contains information that an object ID exists at a location. So, in my setup, LM 1 would send out a vote "object P @ location A". When another Learning Module receives a vote, the vote is transformed so that the receiving Learning Module gets the information that the object ID should exist at the location it is sensing. So, in my setup, LM 2 would receive a vote "object P @ location A" transformed into "object P @ location B". And, at the same time, LM 2 sends out a vote "object P @ location B", which LM 1 would receive transformed into "object P @ location A". So, LM 1 would get reinforcement that object P is at location A from itself and from the vote of LM 2. Also, LM 2 would get reinforcement that object P is at location B from itself and from the LM 1 vote.

The above makes sense to me.

Follow-up question.

Looking at the code, along with the location, we send a "pose" in `morphological_features` (although it seems to me this should maybe be "orientation", since ["pose" == "location" + "orientation"](https://thousandbrainsproject.readme.io/docs/glossary#pose)):

https://github.com/thousandbrainsproject/tbp.monty/blob/fbed64f11bd70eda31fe3e79cb5d945a55ae519d/src/tbp/monty/frameworks/models/evidence_matching/learning_module.py#L373-L390

I notice that we _don't use_ the orientation:

https://github.com/thousandbrainsproject/tbp.monty/blob/fbed64f11bd70eda31fe3e79cb5d945a55ae519d/src/tbp/monty/frameworks/models/evidence_matching/learning_module.py#L864-L866

So, in practice, we only vote on locations, but it is a bit confusing seeing orientations sent out in the votes. Our docs state that we don't do it, while the code does.

I noticed we say "Currently we do not send out pose hypotheses when we are voting" in [Use Pose for Voting improvement doc](https://thousandbrainsproject.readme.io/docs/use-pose-for-voting). Aren't we, though?

nielsleadholm replied:
In that case I think it would be worth updating the description and diagram, which current implies the orientation is used.

Re. longer-term, I think the advantage of applying the rotation is that it would enable similar (but not exactly the same) rotations, or symmetric rotations, to correctly update the location in a similar way. But yeah this is something we can revisit later.

vkakerbeck replied:
No, the diagram shows applying the rotation difference between the two sensory inputs, which we definitely do do here: https://github.com/thousandbrainsproject/tbp.monty/blob/a408bf6063852323b98e009da5e1373d097beb73/src/tbp/monty/frameworks/models/evidence_matching/model.py#L77 (what I also link in this doc)

nielsleadholm reviewed (commented):
**praise:** Beautiful animations and great descriptions.

Just a couple minor comments, and then adding to the discussion around voting.

vkakerbeck commented:
@nielsleadholm I updated the figures with the different colors like we discussed. I also added a new example on voting to show how the rotation hypothesis is applied to the sensor displacement. Let me know what you think.

vkakerbeck commented on docs/figures/how-monty-works/reference_frames_overview.png:
oh good catch! @codeallthethingz did you have some tool for adding white background? Otherwise I'll reexport them

vkakerbeck replied:
nvm, I just updated it :)
